theano/gpuarray/basic_ops.py

```diff
@@ -1613,21 +1611,25 @@ def grad(self, inp, grads):
 
     def gpu_kernels(self, node, name):
         code = """
-KERNEL void eye(GLOBAL_MEM %(ctype)s *a, ga_size n, ga_size m) {
-    ga_size nb = n < m ? n : m;
-    for (ga_size i = LID_0; i < nb; i += LDIM_0) {
-        a[i*m + i] = %(write_a)s(1);
+KERNEL void eye(GLOBAL_MEM %(ctype)s *a, ga_size n, ga_size m, ga_long k) {
+    ga_long col_off = max(k, (ga_long) 0);
+    ga_long row_off = -min(k, (ga_long) 0);
+    if (row_off < n && col_off < m) {
```

Comment:
Instead of doing the check in the kernel, it would probably be a better idea to do this check outside and skip the kernel call completely when the output would be all-zeros.